pkg/registry/template/processor.go

```diff
@@ -0,0 +1,244 @@
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/golang/glog"
+	"github.com/ugorji/go/codec"
+	"strconv"
+)
+
+// TemplateProcessor implements parameter substitution logic for templates
+type TemplateProcessor interface {
+	Process(ctx api.Context, params *extensions.TemplateParameters) (*api.List, error)
+}
+
+type templateProcessor struct {
+	Registry
+	decoder runtime.Decoder
+}
+
+func NewTemplateProcessor(registry Registry, decoder runtime.Decoder) TemplateProcessor {
+	return &templateProcessor{registry, decoder}
+}
+
+var _ TemplateProcessor = &templateProcessor{}
+
+func (tp *templateProcessor) Process(ctx api.Context, params *extensions.TemplateParameters) (*api.List, error) {
+	templateErrors := field.ErrorList{}
+	result := &api.List{}
+
+	// Lookup the Template
+	template, err := tp.GetTemplate(ctx, params.Name)
+	if err != nil {
+		return nil, errors.NewBadRequest(err.Error())
+	}
+
+	// Validate the TemplateParameters are good for the Template
+	if err := ValidateProcessTemplate(params, template); err != nil {
+		return nil, errors.NewBadRequest(err.Error())
+	}
+
+	// Get an object to replace the string values
+	strReplacer, boolReplacer := GetParameterReplacer(params, template)
+
+	// Perform the Substitutions on each Object in the Template
+	for i, obj := range template.Spec.Objects {
+		path := field.NewPath("spec", "objects").Index(i)
+		finishedObj, err := tp.SubstituteParams(path, obj, strReplacer, boolReplacer)
+		if err != nil {
+			templateErrors = append(templateErrors, err)
+			continue
+		}
+
+		result.Items = append(result.Items, finishedObj)
+	}
+	if len(templateErrors) > 0 {
+		return nil, errors.NewInvalid(params.GroupVersionKind().GroupKind(), params.Name, templateErrors)
+	}
+
+	return result, nil
+}
+
+// GetParameterReplacer Returns an object to perform string substitutions
+func GetParameterReplacer(params *extensions.TemplateParameters,
+	template *extensions.Template) (*strings.Replacer, *strings.Replacer) {
+	replaceWith := []string{}
+	boolReplaceWith := []string{}
+	substitutionMap := map[string]extensions.Parameter{}
+	for _, p := range template.Spec.Parameters {
+		substitutionMap[p.Name] = p
+	}
+
+	// Copy the TemplateParameters values into the Template.Parameters structures.
+	// This will override default values when specified.
+	for name, value := range params.ParameterValues {
+		// Doesn't allow updating this field in the struct directly from the map
+		// so we have to copy it out, update, and then re-insert
+		p := substitutionMap[name]
+		p.Value = value
+		substitutionMap[name] = p
+	}
+
+	// Create the Replacer substitution mappings for the Parameters.
+	for k, v := range substitutionMap {
+		// TODO: Figure out if we need to support $((NAME)) syntax since
+		// Replace booleans separately from Strings/Integers so that we can do the appropriate
+		// type conversion.  Ints are automatically type converted for us.
+		if v.Type != extensions.BoolParam {
```

Comment:
v.Type was not intended to be an enforced type or a required value in the parameter definition, it was only supposed to be a hint to users/user interfaces about the intended type.  In fact we don't want it to be enforced explicitly, because I might want to use the same parameter value in multiple ways/as different types:

serviceName: "foosvc_$(PORT)"
servicePort: "$((PORT)""

so that i end up with:
serviceName: "foosvc_8080"
servicePort: 8080